### DISCUSSION TITLE
Shall apollo-client keep supporting java 1.7?

### DISCUSSION THREAD

**Discussion by nobodyiam** [apolloconfig/apollo#3896] [General]
Hi Guys,

I'd like to start this discussion since we met a conflict here.
There was a request to ask apollo-client to support jdk 16(#3642) and that was solved by upgrading guice to 5.0.1(#3646).
However we just found guice stopped supporting jdk 1.7 since [version 4.2.3](https://github.com/google/guice/pull/1240/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R274), which means if we release apollo 1.9.0 now, then apollo-client won't support jdk 1.7 anymore.
I think we now have 2 choices:
1. revert #3646
    * apollo-client 1.9.0 still supports jdk 1.7 but not supports jdk 16
    * we could think about deprecating jdk 1.7 support in the next release
2. keep #3646
    * we need to change the jdk requirement for apollo-client and record this breaking change in release note

Feel free to share your ideas.

**Comment by nobodyiam** [comment]
According to [java-se-support-roadmap](https://www.oracle.com/java/technologies/java-se-support-roadmap.html), jdk 16 is a non-LTS version, so we decided to revert #3646 and discuss whether to support jdk 17(LTS version) when it is released.

**Comment by nobodyiam** [comment]
Java 17 was [released](https://www.oracle.com/news/announcement/oracle-releases-java-17-2021-09-14/) today, so I think it's time to   continue this discussion. Please feel free to share your ideas. cc @apolloconfig/committers

**Comment by kezhenxu94** [comment]
I'm +1 to drop support for 1.7, as for the version, my thought is to release a final `1.10.0` that still support 1.7 with known bug fixed, and then bump up to `2.0` and drop support of 1.7

**Reply by Anilople** [reply]
Though in my company, all Java program use apollo-client is running on jdk8，but I prefer choice 1. 

Most companies won't use jdk16 within 3 years, some companies maybe will still use jdk 1.7 within 2 years.

Drop the support of 1.7 when apollo-client's major version changed from 1 to 2 seems a nice solution.

**Reply by slankka** [reply]
+65535: Prefer choice 1 with no doubt. 

For those companies whose need is JDK1.7, they can choose 1.7 final release. 

Dropping support for JDK1.7 is good for keeping involved with latest technology, and may reduce work for legacy restriction.

**Reply by nobodyiam** [reply]
Looks like this proposal is mostly favored and then we need to see how to achieve this.

> release a final 1.10.0 that still support 1.7 with known bug fixed

Currently 1.9.1 is the latest version that still supports 1.7 and fixes all the known bugs. So we could create a 1.x branch based on 1.9.1 branch and release 1.10.0 later? 

> and then bump up to 2.0 and drop support of 1.7

Then we could update the version of master branch to 2.0.0 and it will introduce some incompatibility changes like updating java version to 1.8 for apollo-client and also other major changes like #3866 which involves complicated conversion steps when upgrading?

**Reply by nisiyong** [reply]
Agree. I think involving this PR #3866 in 2.0.0 is better because they have some steps to upgrade.

**Comment by nisiyong** [comment]
I'm +1 to drop support for 1.7.

For Apollo server-side, JDK8 is fine even though business applications run on JDK7 because the company that deploys Apollo
should prepare individual machines, and they can install JDK8+ on those machines.

We should only care about the client-side. According [Compatibility Guide for JDK 8](https://www.oracle.com/java/technologies/javase/8-compatibility-guide.html)
> Class files built with the Java SE 8 compiler will not run on earlier releases of Java SE.

I think bumping up to `2.0` to support JDK8+ is a good idea. We keep `1.x` to support 1.7.
But if apollo-client will introduce more features in the future, I think we should stop supporting `1.x`.
Maybe we can claim the support end date for `1.x` like the JDK roadmap.

**Reply by kezhenxu94** [reply]
If we cut the final release for 1.x and bump up to 2.0, I don't think we should continue to bring new features to 1.x, which makes it no point to bump up to 2.0.

**Comment by klboke** [comment]
I support the latest version to directly support jdk17. Most java open source projects are also planned like this. For example, kafka3.0 will give up support for jdk8, and spring 6.0 will be developed directly based on jdk17. However, the released version will continue to support jdk7 bug fixes.

**Comment by nobodyiam** [comment]
Based on the discussions, here is my suggestion:
1. Create a new branch `1.x` based on 1.9.1 release
  * `1.x` branch will continue to fix bugs in apollo-client but no features will be introduced
  * `1.x` branch will keep supporting Java 1.7 until there is further decision to deprecate Java 1.7 support
2. Master branch will be versioned to `2.0.0`
  * Master branch will drop Java 1.7 support and add Java 17 support

What do you think? @apolloconfig/committers

**Reply by kezhenxu94** [reply]
if this thread serves as a voting thread. Here is my +1

**Reply by pengweiqhca** [reply]
Very good.

**Reply by nisiyong** [reply]
+1

**Reply by Anilople** [reply]
+1

**Reply by DiegoKrupitza** [reply]
+1
I think this is a good approach.

**Comment by nobodyiam** [comment]
I created branch [1.x](https://github.com/apolloconfig/apollo/tree/1.x) based on tag [1.9.1](https://github.com/apolloconfig/apollo/compare/1.9.1) and also submitted pull request #4015 to drop java 1.7 support and bump version to 2.0.0 for master branch, please help to review.
Java 17 support will be submitted in a separate pr since it might involve quite a lot of changes.

**Reply by nobodyiam** [reply]
#4060 was merged to support java 17
